File: examples/python/singlekey_client.py
"""
SingleKey API Python Client

A complete Python client for the SingleKey Screening API.

Installation:
    pip install requests

Usage:
    from singlekey_client import SingleKeyClient

    client = SingleKeyClient("your_api_token")
    result = client.create_screening(landlord, tenant)
"""

import requests
import time
import logging
from typing import Optional, Dict, Any, List
from dataclasses import dataclass
from enum import Enum

logger = logging.getLogger(__name__)


class Environment(Enum):
    PRODUCTION = "https://platform.singlekey.com"
    SANDBOX = "https://sandbox.singlekey.com"


@dataclass
class Landlord:
    """Landlord information for screening request."""
    first_name: str
    last_name: str
    email: str
    phone: Optional[str] = None
    external_id: Optional[str] = None


@dataclass
class TenantDOB:
    """Tenant date of birth."""
    year: int
    month: int
    day: int


@dataclass
class Tenant:
    """Tenant information for screening request."""
    first_name: str
    last_name: str
    email: str
    phone: str
    dob: TenantDOB
    address: str
    sin: str  # SIN for Canada, SSN for USA
    external_id: Optional[str] = None
    middle_name: Optional[str] = None
    employer: Optional[str] = None
    job_title: Optional[str] = None
    annual_income: Optional[int] = None


@dataclass
class Property:
    """Property information for screening."""
    address: str
    rent: Optional[int] = None
    unit: Optional[str] = None


class SingleKeyError(Exception):
    """Base exception for SingleKey API errors."""
    def __init__(self, message: str, errors: List[str] = None):
        super().__init__(message)
        self.errors = errors or []


class AuthenticationError(SingleKeyError):
    """Raised when authentication fails."""
    pass


class ValidationError(SingleKeyError):
    """Raised when request validation fails."""
    pass


class NotFoundError(SingleKeyError):
    """Raised when resource is not found."""
    pass


class SingleKeyClient:
    """
    SingleKey API client for tenant screening.

    Example:
        client = SingleKeyClient("your_api_token")

        # Create a screening
        result = client.create_screening(
            landlord=Landlord(
                first_name="John",
                last_name="Smith",
                email="john@example.com"
            ),
            tenant=Tenant(
                first_name="Jane",
                last_name="Doe",
                email="jane@example.com",
                phone="5551234567",
                dob=TenantDOB(year=1990, month=6, day=15),
                address="123 Main St, Toronto, ON, Canada, M5V 1A1",
                sin="123456789"
            )
        )

        # Get report
        report = client.get_report(result["purchase_token"])
    """

    def __init__(
        self,
        api_token: str,
        environment: Environment = Environment.PRODUCTION,
        timeout: int = 30
    ):
        """
        Initialize the SingleKey client.

        Args:
            api_token: Your SingleKey API token
            environment: PRODUCTION or SANDBOX
            timeout: Request timeout in seconds
        """
        self.api_token = api_token
        self.base_url = environment.value
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update({
            "Authorization": f"Token {api_token}",
            "Content-Type": "application/json"
        })

    def _request(
        self,
        method: str,
        endpoint: str,
        data: Dict = None,
        params: Dict = None
    ) -> Dict[str, Any]:
        """Make an API request."""
        url = f"{self.base_url}{endpoint}"

        try:
            response = self.session.request(
                method=method,
                url=url,
                json=data,
                params=params,
                timeout=self.timeout
            )

            # Handle different status codes
            if response.status_code == 401:
                raise AuthenticationError("Invalid API token")

            if response.status_code == 404:
                raise NotFoundError("Resource not found")

            result = response.json()

            # Check for API-level errors
            if not result.get("success", True) and result.get("errors"):
                raise ValidationError(
                    result.get("detail", "Validation failed"),
                    result.get("errors", [])
                )

            return result

        except requests.exceptions.Timeout:
            raise SingleKeyError("Request timed out")
        except requests.exceptions.ConnectionError:
            raise SingleKeyError("Connection failed")

    def create_screening(
        self,
        landlord: Landlord,
        tenant: Tenant,
        property_info: Property = None,
        run_now: bool = True,
        tenant_pays: bool = False,
        callback_url: str = None,
        external_deal_id: str = None
    ) -> Dict[str, Any]:
        """
        Create a new screening request.

        Args:
            landlord: Landlord information
            tenant: Tenant information
            property_info: Property details (optional)
            run_now: Process immediately (default True)
            tenant_pays: Tenant provides payment (default False)
            callback_url: Webhook URL for notifications
            external_deal_id: Your CRM deal ID

        Returns:
            Dict with purchase_token and status
        """
        payload = {
            "external_customer_id": landlord.external_id or f"ll-{landlord.email}",
            "external_tenant_id": tenant.external_id or f"tn-{tenant.email}",
            "run_now": run_now,
            "tenant_pays": tenant_pays,

            # Landlord info
            "ll_first_name": landlord.first_name,
            "ll_last_name": landlord.last_name,
            "ll_email": landlord.email,

            # Tenant info
            "ten_first_name": tenant.first_name,
            "ten_last_name": tenant.last_name,
            "ten_email": tenant.email,
            "ten_tel": tenant.phone,
            "ten_dob_year": tenant.dob.year,
            "ten_dob_month": tenant.dob.month,
            "ten_dob_day": tenant.dob.day,
            "ten_address": tenant.address,
            "ten_sin": tenant.sin,
        }

        # Optional fields
        if landlord.phone:
            payload["ll_tel"] = landlord.phone

        if tenant.middle_name:
            payload["ten_middle_name"] = tenant.middle_name

        if tenant.employer:
            payload["ten_employer"] = tenant.employer

        if tenant.job_title:
            payload["ten_job_title"] = tenant.job_title

        if tenant.annual_income:
            payload["ten_annual_income"] = tenant.annual_income

        if property_info:
            payload["purchase_address"] = property_info.address
            if property_info.rent:
                payload["purchase_rent"] = property_info.rent
            if property_info.unit:
                payload["purchase_unit"] = property_info.unit

        if callback_url:
            payload["callback_url"] = callback_url

        if external_deal_id:
            payload["external_deal_id"] = external_deal_id

        return self._request("POST", "/api/request", data=payload)

    def create_form_request(
        self,
        landlord: Landlord,
        tenant_email: str,
        tenant_first_name: str = None,
        tenant_last_name: str = None,
        property_address: str = None,
        tenant_form: bool = False,
        callback_url: str = None
    ) -> Dict[str, Any]:
        """
        Create a form-based screening request.

        Args:
            landlord: Landlord information
            tenant_email: Tenant's email address
            tenant_first_name: Tenant's first name (optional)
            tenant_last_name: Tenant's last name (optional)
            property_address: Property address (required for tenant_form)
            tenant_form: Use direct tenant form (default False)
            callback_url: Webhook URL for notifications

        Returns:
            Dict with purchase_token and form_url
        """
        payload = {
            "external_customer_id": landlord.external_id or f"ll-{landlord.email}",
            "external_tenant_id": f"tn-{tenant_email}",

            # Landlord info
            "ll_first_name": landlord.first_name,
            "ll_last_name": landlord.last_name,
            "ll_email": landlord.email,

            # Tenant info
            "ten_email": tenant_email,
        }

        if landlord.phone:
            payload["ll_tel"] = landlord.phone

        if tenant_first_name:
            payload["ten_first_name"] = tenant_first_name

        if tenant_last_name:
            payload["ten_last_name"] = tenant_last_name

        if tenant_form:
            payload["tenant_form"] = True
            if property_address:
                payload["purchase_address"] = property_address

        if callback_url:
            payload["callback_url"] = callback_url

        return self._request("POST", "/api/request", data=payload)

    def get_report(self, purchase_token: str) -> Dict[str, Any]:
        """
        Get screening report.

        Args:
            purchase_token: Token from create_screening

        Returns:
            Dict with report data or status
        """
        return self._request("GET", f"/api/report/{purchase_token}")

    def get_applicant(
        self,
        purchase_token: str,
        detailed: bool = False,
        show_credit_score: bool = False
    ) -> Dict[str, Any]:
        """
        Get applicant information.

        Args:
            purchase_token: Token from create_screening
            detailed: Include detailed information
            show_credit_score: Include credit score

        Returns:
            Dict with applicant data
        """
        params = {}
        if detailed:
            params["detailed"] = "true"
        if show_credit_score:
            params["show_credit_score"] = "true"

        return self._request(
            "GET",
            f"/api/applicant/{purchase_token}",
            params=params
        )

    def validate_screening(self, screening_id: str) -> Dict[str, Any]:
        """
        Validate screening data for errors.

        Args:
            screening_id: The screening ID to validate

        Returns:
            Dict with validation status and any errors
        """
        return self._request("POST", f"/api/purchase_errors/{screening_id}")

    def download_pdf(self, purchase_token: str, output_path: str) -> bool:
        """
        Download screening report as PDF.

        Args:
            purchase_token: Token from create_screening
            output_path: File path to save the PDF

        Returns:
            True if successful
        """
        url = f"{self.base_url}/api/report_pdf/{purchase_token}"

        response = self.session.get(url, stream=True, timeout=self.timeout)

        if response.status_code == 200:
            with open(output_path, 'wb') as f:
                for chunk in response.iter_content(chunk_size=8192):
                    f.write(chunk)
            return True

        if response.status_code == 404:
            raise NotFoundError("Report not found")

        raise SingleKeyError(f"Failed to download PDF: {response.status_code}")

    def wait_for_report(
        self,
        purchase_token: str,
        timeout: int = 300,
        poll_interval: int = 10
    ) -> Dict[str, Any]:
        """
        Poll until report is ready.

        Args:
            purchase_token: Token from create_screening
            timeout: Maximum seconds to wait
            poll_interval: Seconds between checks

        Returns:
            Completed report data
        """
        start_time = time.time()

        while time.time() - start_time < timeout:
            report = self.get_report(purchase_token)

            if report.get("success") and report.get("singlekey_score"):
                return report

            logger.info(f"Status: {report.get('detail', 'Processing...')}")
            time.sleep(poll_interval)

        raise SingleKeyError("Timeout waiting for report")


# Example usage
if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)

    # Initialize client
    client = SingleKeyClient(
        api_token="your_api_token",
        environment=Environment.SANDBOX
    )

    try:
        # Create screening
        result = client.create_screening(
            landlord=Landlord(
                external_id="landlord-123",
                first_name="John",
                last_name="Smith",
                email="john@example.com",
                phone="5551234567"
            ),
            tenant=Tenant(
                external_id="tenant-456",
                first_name="Jane",
                last_name="Doe",
                email="jane@example.com",
                phone="5559876543",
                dob=TenantDOB(year=1990, month=6, day=15),
                address="456 Oak Ave, Toronto, ON, Canada, M5V 2B3",
                sin="123456789"
            ),
            property_info=Property(
                address="123 Main St, Toronto, ON, Canada, M5V 1A1",
                rent=2000
            ),
            callback_url="https://yoursite.com/webhooks/singlekey"
        )

        print(f"Screening created: {result['purchase_token']}")

        # Wait for report
        report = client.wait_for_report(result['purchase_token'])

        print(f"SingleKey Score: {report['singlekey_score']}")
        print(f"Report URL: {report['report_url']}")

        # Download PDF
        client.download_pdf(result['purchase_token'], "screening_report.pdf")
        print("PDF saved to screening_report.pdf")

    except ValidationError as e:
        print(f"Validation failed: {e}")
        for error in e.errors:
            print(f"  - {error}")

    except SingleKeyError as e:
        print(f"API error: {e}")

File: examples/javascript/singlekey-client.js
/**
 * SingleKey API JavaScript/Node.js Client
 *
 * A complete JavaScript client for the SingleKey Screening API.
 *
 * Installation:
 *   npm install axios
 *
 * Usage:
 *   const SingleKeyClient = require('./singlekey-client');
 *   const client = new SingleKeyClient('your_api_token');
 */

const axios = require('axios');

// Environment constants
const Environment = {
  PRODUCTION: 'https://platform.singlekey.com',
  SANDBOX: 'https://sandbox.singlekey.com'
};

// Custom error classes
class SingleKeyError extends Error {
  constructor(message, errors = []) {
    super(message);
    this.name = 'SingleKeyError';
    this.errors = errors;
  }
}

class AuthenticationError extends SingleKeyError {
  constructor(message) {
    super(message);
    this.name = 'AuthenticationError';
  }
}

class ValidationError extends SingleKeyError {
  constructor(message, errors) {
    super(message, errors);
    this.name = 'ValidationError';
  }
}

class NotFoundError extends SingleKeyError {
  constructor(message) {
    super(message);
    this.name = 'NotFoundError';
  }
}

/**
 * SingleKey API Client
 *
 * @example
 * const client = new SingleKeyClient('your_api_token');
 *
 * const result = await client.createScreening({
 *   landlord: {
 *     firstName: 'John',
 *     lastName: 'Smith',
 *     email: 'john@example.com'
 *   },
 *   tenant: {
 *     firstName: 'Jane',
 *     lastName: 'Doe',
 *     email: 'jane@example.com',
 *     phone: '5551234567',
 *     dob: { year: 1990, month: 6, day: 15 },
 *     address: '123 Main St, Toronto, ON, Canada, M5V 1A1',
 *     sin: '123456789'
 *   }
 * });
 */
class SingleKeyClient {
  /**
   * Create a new SingleKey client
   * @param {string} apiToken - Your API token
   * @param {Object} options - Configuration options
   * @param {string} options.environment - PRODUCTION or SANDBOX
   * @param {number} options.timeout - Request timeout in ms
   */
  constructor(apiToken, options = {}) {
    this.apiToken = apiToken;
    this.baseUrl = options.environment || Environment.PRODUCTION;
    this.timeout = options.timeout || 30000;

    this.client = axios.create({
      baseURL: this.baseUrl,
      timeout: this.timeout,
      headers: {
        'Authorization': `Token ${apiToken}`,
        'Content-Type': 'application/json'
      }
    });
  }

  /**
   * Make an API request
   * @private
   */
  async _request(method, endpoint, data = null, params = null) {
    try {
      const response = await this.client.request({
        method,
        url: endpoint,
        data,
        params
      });

      const result = response.data;

      // Check for API-level errors
      if (result.success === false && result.errors) {
        throw new ValidationError(
          result.detail || 'Validation failed',
          result.errors
        );
      }

      return result;

    } catch (error) {
      if (error instanceof SingleKeyError) {
        throw error;
      }

      if (error.response) {
        const status = error.response.status;

        if (status === 401) {
          throw new AuthenticationError('Invalid API token');
        }

        if (status === 404) {
          throw new NotFoundError('Resource not found');
        }

        const data = error.response.data;
        if (data && data.errors) {
          throw new ValidationError(data.detail || 'Request failed', data.errors);
        }

        throw new SingleKeyError(`Request failed with status ${status}`);
      }

      if (error.code === 'ECONNABORTED') {
        throw new SingleKeyError('Request timed out');
      }

      throw new SingleKeyError(error.message);
    }
  }

  /**
   * Create a new screening request
   * @param {Object} options - Screening options
   * @param {Object} options.landlord - Landlord information
   * @param {Object} options.tenant - Tenant information
   * @param {Object} options.property - Property information (optional)
   * @param {boolean} options.runNow - Process immediately (default true)
   * @param {boolean} options.tenantPays - Tenant provides payment
   * @param {string} options.callbackUrl - Webhook URL
   * @returns {Promise<Object>} Screening result with purchase_token
   */
  async createScreening({
    landlord,
    tenant,
    property = null,
    runNow = true,
    tenantPays = false,
    callbackUrl = null,
    externalDealId = null
  }) {
    const payload = {
      external_customer_id: landlord.externalId || `ll-${landlord.email}`,
      external_tenant_id: tenant.externalId || `tn-${tenant.email}`,
      run_now: runNow,
      tenant_pays: tenantPays,

      // Landlord info
      ll_first_name: landlord.firstName,
      ll_last_name: landlord.lastName,
      ll_email: landlord.email,

      // Tenant info
      ten_first_name: tenant.firstName,
      ten_last_name: tenant.lastName,
      ten_email: tenant.email,
      ten_tel: tenant.phone,
      ten_dob_year: tenant.dob.year,
      ten_dob_month: tenant.dob.month,
      ten_dob_day: tenant.dob.day,
      ten_address: tenant.address,
      ten_sin: tenant.sin
    };

    // Optional fields
    if (landlord.phone) payload.ll_tel = landlord.phone;
    if (tenant.middleName) payload.ten_middle_name = tenant.middleName;
    if (tenant.employer) payload.ten_employer = tenant.employer;
    if (tenant.jobTitle) payload.ten_job_title = tenant.jobTitle;
    if (tenant.annualIncome) payload.ten_annual_income = tenant.annualIncome;

    if (property) {
      payload.purchase_address = property.address;
      if (property.rent) payload.purchase_rent = property.rent;
      if (property.unit) payload.purchase_unit = property.unit;
    }

    if (callbackUrl) payload.callback_url = callbackUrl;
    if (externalDealId) payload.external_deal_id = externalDealId;

    return this._request('POST', '/api/request', payload);
  }

  /**
   * Create a form-based screening request
   * @param {Object} options - Form request options
   * @returns {Promise<Object>} Result with form_url
   */
  async createFormRequest({
    landlord,
    tenantEmail,
    tenantFirstName = null,
    tenantLastName = null,
    propertyAddress = null,
    tenantForm = false,
    callbackUrl = null
  }) {
    const payload = {
      external_customer_id: landlord.externalId || `ll-${landlord.email}`,
      external_tenant_id: `tn-${tenantEmail}`,

      ll_first_name: landlord.firstName,
      ll_last_name: landlord.lastName,
      ll_email: landlord.email,

      ten_email: tenantEmail
    };

    if (landlord.phone) payload.ll_tel = landlord.phone;
    if (tenantFirstName) payload.ten_first_name = tenantFirstName;
    if (tenantLastName) payload.ten_last_name = tenantLastName;

    if (tenantForm) {
      payload.tenant_form = true;
      if (propertyAddress) payload.purchase_address = propertyAddress;
    }

    if (callbackUrl) payload.callback_url = callbackUrl;

    return this._request('POST', '/api/request', payload);
  }

  /**
   * Get screening report
   * @param {string} purchaseToken - Token from createScreening
   * @returns {Promise<Object>} Report data or status
   */
  async getReport(purchaseToken) {
    return this._request('GET', `/api/report/${purchaseToken}`);
  }

  /**
   * Get applicant information
   * @param {string} purchaseToken - Token from createScreening
   * @param {Object} options - Query options
   * @returns {Promise<Object>} Applicant data
   */
  async getApplicant(purchaseToken, { detailed = false, showCreditScore = false } = {}) {
    const params = {};
    if (detailed) params.detailed = 'true';
    if (showCreditScore) params.show_credit_score = 'true';

    return this._request('GET', `/api/applicant/${purchaseToken}`, null, params);
  }

  /**
   * Validate screening for errors
   * @param {string} screeningId - Screening ID to validate
   * @returns {Promise<Object>} Validation result
   */
  async validateScreening(screeningId) {
    return this._request('POST', `/api/purchase_errors/${screeningId}`);
  }

  /**
   * Download report as PDF
   * @param {string} purchaseToken - Token from createScreening
   * @returns {Promise<Buffer>} PDF file buffer
   */
  async downloadPdf(purchaseToken) {
    const response = await this.client.get(`/api/report_pdf/${purchaseToken}`, {
      responseType: 'arraybuffer'
    });

    return Buffer.from(response.data);
  }

  /**
   * Poll until report is ready
   * @param {string} purchaseToken - Token from createScreening
   * @param {Object} options - Polling options
   * @returns {Promise<Object>} Completed report
   */
  async waitForReport(purchaseToken, { timeout = 300000, pollInterval = 10000 } = {}) {
    const startTime = Date.now();

    while (Date.now() - startTime < timeout) {
      const report = await this.getReport(purchaseToken);

      if (report.success && report.singlekey_score) {
        return report;
      }

      console.log(`Status: ${report.detail || 'Processing...'}`);
      await new Promise(resolve => setTimeout(resolve, pollInterval));
    }

    throw new SingleKeyError('Timeout waiting for report');
  }
}

// Export for Node.js
module.exports = {
  SingleKeyClient,
  SingleKeyError,
  AuthenticationError,
  ValidationError,
  NotFoundError,
  Environment
};

// Example usage
async function main() {
  const client = new SingleKeyClient('your_api_token', {
    environment: Environment.SANDBOX
  });

  try {
    // Create screening
    const result = await client.createScreening({
      landlord: {
        externalId: 'landlord-123',
        firstName: 'John',
        lastName: 'Smith',
        email: 'john@example.com',
        phone: '5551234567'
      },
      tenant: {
        externalId: 'tenant-456',
        firstName: 'Jane',
        lastName: 'Doe',
        email: 'jane@example.com',
        phone: '5559876543',
        dob: { year: 1990, month: 6, day: 15 },
        address: '456 Oak Ave, Toronto, ON, Canada, M5V 2B3',
        sin: '123456789'
      },
      property: {
        address: '123 Main St, Toronto, ON, Canada, M5V 1A1',
        rent: 2000
      },
      callbackUrl: 'https://yoursite.com/webhooks/singlekey'
    });

    console.log(`Screening created: ${result.purchase_token}`);

    // Wait for report
    const report = await client.waitForReport(result.purchase_token);

    console.log(`SingleKey Score: ${report.singlekey_score}`);
    console.log(`Report URL: ${report.report_url}`);

    // Download PDF
    const fs = require('fs');
    const pdfBuffer = await client.downloadPdf(result.purchase_token);
    fs.writeFileSync('screening_report.pdf', pdfBuffer);
    console.log('PDF saved to screening_report.pdf');

  } catch (error) {
    if (error instanceof ValidationError) {
      console.error('Validation failed:', error.message);
      error.errors.forEach(err => console.error(`  - ${err}`));
    } else {
      console.error('API error:', error.message);
    }
  }
}

// Run example if executed directly
if (require.main === module) {
  main();
}
